### PR TITLE
feat(efb): tabbed settings page and new autobrightness option

### LINKED ISSUE #4865: EFB Settings view layout update
### Aircraft Version

Development

### Description

Currently, the EFB settings view is running out of room for additional settings - as EFB evolves it's going to need the ability to customize/tailor it and I am already seeing additional inputs being required.

Suggesting the separation of EFB settings into views based on their current grouping, perhaps with a tabs-based approach exactly like what is being used in the Dispatch view.

### References (optional)

image

### Additional info (optional)

_No response_

### Discord Username (optional)

antiso#8258

### PR DESCRIPTION
**Pull request by ErickSharp**
Fixes #4865

## Summary of Changes

Adds a tabbed settings page to the efb. In addition to a brand new EFB autobrightness option.

## Screenshots (if necessary)

image

## Additional context

Discord username (if different from GitHub): Erick [Z-6]#6484

## Testing instructions

Play around with the tabs. At the moment, because there is no more than one page per tab, the arrow buttons at the bottom won't take you anywhere, but they do work... I promise!

Turn on the autobrightness option and change the time using the weather slider. See the brightness of the display change and the brightness slider should be locked. 

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/instruments/src/EFB/Settings/Settings.tsx [hunk 1]
```diff
@@ -103,6 +65,127 @@ const AircraftConfigurationPage = () => {
         }
     };
 
+    return (
+        <div className="bg-navy-lighter rounded-xl px-6 shadow-lg divide-y divide-gray-700 flex flex-col">
+
+            <div className="py-4 flex flex-row justify-between items-center">
+                <span className="text-lg text-gray-300">Thrust Reduction Altitude (ft)</span>
```
Comment (tracernz) [inline src/instruments/src/EFB/Settings/Settings.tsx:72]: These should be heights, not altitudes. The FMC adds these heights to the departure runway altitude to get the altitude... e.g. default thrust reduction alt = (departure runway altitude) + (default thrust reduction height setting)

File: src/instruments/src/EFB/Settings/Settings.tsx [hunk 2]
```diff
@@ -103,6 +65,127 @@ const AircraftConfigurationPage = () => {
...
+                            border-2 border-navy-light focus-within:outline-none focus-within:border-teal-light-contrast text-center"
+                        placeholder={accelerationAlt}
+                        noLabel
+                        value={accelerationAltSetting}
+                        onChange={(event) => handleSetAccelerationAlt(event)}
+                    />
+                </div>
+            </div>
+            <div className="py-4 flex flex-row justify-between items-center">
+                <span className="text-lg text-gray-300">Acceleration Out Altitude (ft)</span>
```
Comment (tracernz) [inline src/instruments/src/EFB/Settings/Settings.tsx:98]: Should be "Engine Out Acceleration Height"

File: src/instruments/src/EFB/Settings/Settings.tsx [hunk 3]
```diff
@@ -103,6 +65,127 @@ const AircraftConfigurationPage = () => {
...
+                            border-2 border-navy-light focus-within:outline-none focus-within:border-teal-light-contrast text-center"
+                        placeholder={thrustReductionAlt}
+                        noLabel
+                        value={thrustReductionAltSetting}
+                        onChange={(event) => handleSetThrustReductionAlt(event)}
+                    />
+                </div>
+            </div>
+            <div className="py-4 flex flex-row justify-between items-center">
+                <span className="text-lg text-gray-300">Acceleration Altitude (ft)</span>
```
Comment (tracernz) [inline src/instruments/src/EFB/Settings/Settings.tsx:85]: As above

File: src/instruments/src/EFB/Settings/Settings.tsx [hunk 4]
```diff
@@ -16,386 +18,448 @@ type AdirsButton = {
...
+        </div>
+    );
+};
+
+const SimOptionsPage = () => {
+    const [showThrottleSettings, setShowThrottleSettings] = useState(false);
+    const { setShowNavbar } = useContext(SettingsNavbarContext);
+
+    const [adirsAlignTime, setAdirsAlignTime] = usePersistentProperty('CONFIG_ALIGN_TIME', 'REAL');
+    const [_, setAdirsAlignTimeSimVar] = useSimVar('L:A32NX_CONFIG_ADIRS_IR_ALIGN_TIME', 'Enum', Number.MAX_SAFE_INTEGER);
```
Comment (2hwk) [inline src/instruments/src/EFB/Settings/Settings.tsx:166]: The convention here for useSimVar is ``[, setSimVar]`` not ``[_, setSimVar]``. I know it was like this in the original code too but may as well correct this.

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: src/instruments/src/EFB/Settings/Settings.tsx (hunk 4 above)
Comment (ErickSharp) [reply src/instruments/src/EFB/Settings/Settings.tsx:166]: thanks for letting me know : )

**Review by 2hwk — APPROVED** [review]

File: src/instruments/src/EFB/Efb.tsx [hunk 5]
```diff
@@ -118,9 +119,23 @@ const navigraph = new NavigraphClient();
...
+    const [currentLocalTime] = useSimVar('E:LOCAL TIME', 'seconds', 3000);
+    const [, setBrightness] = useSimVar('L:A32NX_EFB_BRIGHTNESS', 'number');
+    const [usingAutobrightness] = useSimVar('L:A32NX_EFB_USING_AUTOBRIGHTNESS', 'bool', 5000);
+
+    // handle setting brightness if user is using autobrightness
+    useEffect(() => {
+        if (usingAutobrightness) {
+            const localTime = currentLocalTime / 3600;
+            // eslint-disable-next-line no-restricted-properties
+            setBrightness(Math.pow(Math.E, -(Math.pow(localTime - 12, 2) / 24)) * 100);
```
Comment (MisterChocker) [inline src/instruments/src/EFB/Efb.tsx:131]: interesting how we use three different approaches across the plane to automatically adjust the display brightness.

**Comment by St54Kevin** [comment]
**Quality Assurance Tester**

Object of testing: #5524
Tier of Testing: 1
Date: 14/08/2021

**Testing Process:**
1. Played with the settings tabs and changed some settings
2. enabled Auto brightness and played with time of day

**Negatives:** 
- I dont like that the autobrightness gets adjusted around the whole day. The sun is up from sunrise until sunset so between that it should be 100% brightness in my opinion. Now its 100% at 12 noon and then gets decreased. But you may need 100% brightness even 8-10 hours later 

**Testing Results:** Partially Passed

**Comment by Benjozork** [comment]
Shouldn't we specifically refer to OPC/AMI and pin programs ?

**Comment by NUTTTTTTTT** [comment]
Quality Assurance Tester Report

Discord: █▀█ █▄█ ▀█▀#2123
Object of testing: #PR 5524
Tier of Testing: 1
Date: 14/08/2021

Testing Process:
- Loaded in on runway at EGSS, turned on EFB, 
- Messed around between tabs - all worked as expected
- Turned on auto brightness, changed times of day and weather conditions to adjust lighting

Negatives: 
- Auto-brightness seems to be overly dark in some cases even when still bright in the evenings and kind of hard to read, obviously can be switched off in the flyPad tab of the settings but in general I feel the auto-brightness needs kicking up a notch. 
- No adjustment when in stormy conditions and cloud coverage is blocking light.

Testing Results: Partial-Pass

Conclusions:
- The tabbed settings pages work great, although a bit bare at the moment, these will be a massive improvement as time goes on and new settings are added under different categories, and will make for improved ease of use, as well as making it easier to direct users to a setting in #support cases.

- The auto-brightness is definitely a cool feature, but I feel needs some tweaking before it's an improvement of functionality, I'd personally need to keep this off in my general use. 

Media:
image
image

**Comment by NUTTTTTTTT** [comment]
Quality Assurance Tester Report

Discord: █▀█ █▄█ ▀█▀#2123
Object of testing: #PR 5638
Tier of Testing: 1
Date: 14/08/2021

Testing Process:
- Loaded in on runway at EGSS.
- Tested ECAM buttons and checked colours. - Passed
- Checked FCU buttons to ensure they remained original colour (Green). - Passed
- Tested ANN LT TEST - checked over all lights to check they looked the same as before. - Passed

Negatives: N/A
Testing Results: Pass

Conclusions: Works exactly as intended, a nice improvement and looks very clean! Great stuff!


Media:
With ANN LT TEST ON 
image
image

**Comment by ErickSharp** [comment]
> Shouldn't we specifically refer to OPC/AMI and pin programs in the UI ?

Mike gave me an MD file with the titles and contents of each settings page and that’s what I put in, maybe you would like to discuss with him?

File: src/instruments/src/EFB/Efb.tsx (hunk 5 above)
Comment (ErickSharp) [reply src/instruments/src/EFB/Efb.tsx:131]: For all I know, other solutions are based on breakpoints of either day/night or a couple more in between. This approach uses a bell curve function which I think would much better match the functionality of a sensor that detects ambient light intensity.

**Comment by ErickSharp** [comment]
@NUTTTTTTTT @St54Kevin Thank you both for your feedback. I took it, experimented and have chosen a different calculation. This still ignored the latitude of where you are, but I believe it to be drastically better than what you guys initially tested. Please try it out!

**Comment by NUTTTTTTTT** [comment]
Quality Assurance Tester Report

Discord: █▀█ █▄█ ▀█▀#2123
Object of testing: #PR 5524
Tier of Testing: 1
Date: 15/08/2021

Follow up report following tweaks to auto-brightness
Testing Process:
- Loaded in on runway at EGSS, turned on EFB,
- Turned auto-brightness on and messed with time of day

Conclusions:
- Auto-brightness stays brighter for longer and much more usable in evening/early morning conditions, would genuinely use this compared to the previous settings for the PR. A nice addition now!

Result: PASS

Media:
BEFORE TWEAKING
image

AFTER TWEAKING
image

**Comment by tracernz** [comment]
> Shouldn't we specifically refer to OPC/AMI and pin programs in the UI ?

What would we do for things like weight unit, where it is both?

**Comment by FlightSimmer7700** [comment]
Quality Assurance Tester

Discord: FlightSimmer7700#3468
Object of testing: #5524 
Tier of Testing: 1
Date: 15/8/21

Testing Process: 
Loaded in at an airport in Africa and one in a very northerly area of the world. Looked through pages and checked the auto-brightness
Negatives: 
- When auto-brightness sets itself to 0 it can be bugged and starts to move up and down constantly (the slider)
- Because this system relies on time to set the brightness it can be awkward for places at the equator and very north/south poles where the extremes of light are experienced. This will lead to it being mid-day and it being fully dark but the efb will be at full brightness and opposite at the opposite pole
- When calibrating and pressing back, at the top the "Defaults" page will be highlighted but the page will actually be the "sim options" one instead and needs to be reset by clicking one of the pages at the top

Testing Results: Not passed

Conclusions: Pages and auto-brightness work under most normal conditions but have some bugs that need ironing out
Media: 
Brightness moves up and down constantly 
https://user-images.githubusercontent.com/70103461/129464171-471352a0-25e9-4b0a-8cbc-02903bf35afe.mp4
image

Dark but full brightness:
https://user-images.githubusercontent.com/70103461/129464260-bb81e354-c77c-4276-af52-aa6c2ed9b4b2.mp4

Calibration page bug:
https://user-images.githubusercontent.com/70103461/129464265-54824445-6f4e-43b7-865d-4c891844381b.mp4

**Comment by ErickSharp** [comment]
> Quality Assurance Tester
> 
> Discord: FlightSimmer7700#3468
> Object of testing: #5524
> Tier of Testing: 1
> Date: 15/8/21
> 
> Testing Process:
> Loaded in at an airport in Africa and one in a very northerly area of the world. Looked through pages and checked the auto-brightness
> Negatives:
> 
> * When auto-brightness sets itself to 0 it can be bugged and starts to move up and down constantly (the slider)
> * Because this system relies on time to set the brightness it can be awkward for places at the equator and very north/south poles where the extremes of light are experienced. This will lead to it being mid-day and it being fully dark but the efb will be at full brightness and opposite at the opposite pole
> * When calibrating and pressing back, at the top the "Defaults" page will be highlighted but the page will actually be the "sim options" one instead and needs to be reset by clicking one of the pages at the top
> 
> Testing Results: Not passed
> 
> Conclusions: Pages and auto-brightness work under most normal conditions but have some bugs that need ironing out
> Media:
> Brightness moves up and down constantly
> https://user-images.githubusercontent.com/70103461/129464171-471352a0-25e9-4b0a-8cbc-02903bf35afe.mp4
> 
> Dark but full brightness:
> https://user-images.githubusercontent.com/70103461/129464260-bb81e354-c77c-4276-af52-aa6c2ed9b4b2.mp4
> 
> Calibration page bug:
> https://user-images.githubusercontent.com/70103461/129464265-54824445-6f4e-43b7-865d-4c891844381b.mp4

The first and third issues have been fixed, however a calculation that always works appropriately needs to be dependant on both your latitude and the season/date. That is more complex and I would have to address that at a later time.

**Comment by MisterChocker** [comment]
@ErickSharp take a look at how the isis and clock do it, they work on the similar bases except the isis is a bit more advanced since user can "manipulate" it.

**Comment by FlightSimmer7700** [comment]
> > Quality Assurance Tester
> > Discord: FlightSimmer7700#3468
> > Object of testing: #5524
> > Tier of Testing: 1
> > Date: 15/8/21
> > Testing Process:
> > Loaded in at an airport in Africa and one in a very northerly area of the world. Looked through pages and checked the auto-brightness
> > Negatives:
> > 
> > * When auto-brightness sets itself to 0 it can be bugged and starts to move up and down constantly (the slider)
> > * Because this system relies on time to set the brightness it can be awkward for places at the equator and very north/south poles where the extremes of light are experienced. This will lead to it being mid-day and it being fully dark but the efb will be at full brightness and opposite at the opposite pole
> > * When calibrating and pressing back, at the top the "Defaults" page will be highlighted but the page will actually be the "sim options" one instead and needs to be reset by clicking one of the pages at the top
> > 
> > Testing Results: Not passed
> > Conclusions: Pages and auto-brightness work under most normal conditions but have some bugs that need ironing out
> > Media:
> > Brightness moves up and down constantly
> > https://user-images.githubusercontent.com/70103461/129464171-471352a0-25e9-4b0a-8cbc-02903bf35afe.mp4
> > Dark but full brightness:
> > https://user-images.githubusercontent.com/70103461/129464260-bb81e354-c77c-4276-af52-aa6c2ed9b4b2.mp4
> > Calibration page bug:
> > https://user-images.githubusercontent.com/70103461/129464265-54824445-6f4e-43b7-865d-4c891844381b.mp4
> 
> The first and third issues have been fixed, however a calculation that always works appropriately in all parts of the world at all dates and times needs to be a calculation that provides brightness output as a function of date, time of day, latitude, and maybe even aerosol density. That is far more complex and I would have to address that at a later time. For the time being, however, I think we can go forward with merging this PR.

Fixed issues tested and work well, would prefer if the third issue also got fixed under this pr as it is being implemented and so then won't need a second pr just to fix that but up to you I guess

**Pull request merged**